--- rt2500-1.1.0-b4/Module/rtmp_main.c.	2008-03-02 23:07:01.962750249 +0100
+++ rt2500-1.1.0-b4/Module/rtmp_main.c	2008-03-02 23:07:55.959039225 +0100
@@ -199,8 +199,9 @@ INT __devinit   RT2500_probe(
         DBGPRINT(RT_DEBUG_TRACE, "init_ethernet failed\n");
         goto err_out;
     }
-
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,24)
     SET_MODULE_OWNER(net_dev);
+#endif
 
     #if (LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,0))
        SET_NETDEV_DEV(net_dev, &(pPci_Dev->dev));
@@ -339,8 +340,11 @@ INT RT2500_open(
     {
         NICDisableInterrupt(pAd);
     }
-
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,24)
+    status = request_irq(pAd->pPci_Dev->irq, &RTMPIsr, IRQF_SHARED, net_dev->name, net_dev);
+#else
     status = request_irq(pAd->pPci_Dev->irq, &RTMPIsr, SA_SHIRQ, net_dev->name, net_dev);
+#endif
     if (status)
     {
         goto out_module_put;
